host/src/proof.rs

```diff
@@ -256,10 +268,20 @@ impl ProofActor {
                         error!("Failed to cancel task: {error}")
                     }
                 }
-                Message::Task(proof_request) => {
+                Message::Task(mut proof_request) => {
                     debug!("Message::Task({proof_request:?})");
                     let running_task_count = self.running_tasks.lock().await.len();
                     if running_task_count < self.opts.concurrency_limit {
+                        // Set gpu number for task
+                        if proof_request.gpu_number.is_some() {  panic!("GPU number is already set"); }
+                        let mut available_gpus = self.available_gpus.lock().await;
+                        if let Some(gpu_number) = available_gpus.pop_front() {
+                            proof_request.set_gpu_number(Some(gpu_number));
+                        } else {
+                            panic!("No available GPU");
+                        }
+                        drop(available_gpus);
+
```

Comment:
Maybe this? It uses a closure so no need for an explicit drop too. Not a Rust expert though, so must be taken with caution!
```suggestion
                        proof_request.set_gpu_number(Some({
                            proof_request.gpu_number
                                .as_ref()
                                .inspect(|_| panic!("GPU number is already set"));
                        
                            self.available_gpus
                                .lock()
                                .await
                                .pop_front()
                                .ok_or_else(|| panic!("No available GPU"))
                        }));
```